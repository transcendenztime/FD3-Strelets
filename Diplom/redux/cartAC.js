const QTY_ADD = 'QTY_ADD';
const PROD_ADD = 'PROD_ADD';
const PROD_DEL = 'PROD_DEL';
const SUM_QTY = 'SUM_QTY';
const QTY_TO_NULL = 'QTY_TO_NULL';
const DELIVERY_CHNG = 'DELIVERY_CHNG';

const CART_CLEAR = 'CART_CLEAR';

const prod_add = function(prodid, info) {
  return {
    type: PROD_ADD,
    prodid: prodid,
    info: info,
  };
}

const prod_del = function(prodid) {
  return {
    type: PROD_DEL,
    prodid: prodid,
  };
}

const cart_clear = function() {
  return {
    type: CART_CLEAR,
  };
}

const qty_add = function(prodid, newQTY, newSum) {
  return {
    type: QTY_ADD,
    prodid: prodid,
    newQTY: newQTY,
    newSum: newSum
  };
}

const sum_qty = function(qty) {
  return {
    type: SUM_QTY,
    qty: qty,
  }
}

const qty_to_null = function() {
  return {
    type: QTY_TO_NULL,
  }
}

const delivery_chng = function(num) {
  return {
    type: DELIVERY_CHNG,
    num: num,
  }
}

export {
  prod_add, PROD_ADD,
  prod_del, PROD_DEL,
  cart_clear, CART_CLEAR,
  qty_add, QTY_ADD,
  sum_qty, SUM_QTY,
  qty_to_null, QTY_TO_NULL,
  delivery_chng, DELIVERY_CHNG,
}
