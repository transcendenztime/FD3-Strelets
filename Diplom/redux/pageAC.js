const PAGE_CHANGE = 'PAGE_CHANGE';

const page_change = function(page) {
  return {
    type: PAGE_CHANGE,
    page: page,
  };
}

export {
  page_change, PAGE_CHANGE,
}
