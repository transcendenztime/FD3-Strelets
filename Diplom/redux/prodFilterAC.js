const PROD_FILTER = 'PROD_FILTER';

const prod_filter = function(name) {
  return {
    type: PROD_FILTER,
    name,
  };
}

export {
    prod_filter, PROD_FILTER,
}