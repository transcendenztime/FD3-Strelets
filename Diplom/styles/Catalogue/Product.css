.products-container {
    display: flex;
    justify-content: start;
    flex-wrap: wrap;
}

.product-wrap {
    width: 30%;
    box-sizing: border-box;
    margin: 5px;
    padding: 10px;
    text-align: center;
    border: 1px solid #f5f5f5;
}

.product-title {
    display: inline-block;
    color: black;
    text-decoration: none;
    height: 60px;
}
.product-img-container {    
    width: 80%;
    height: 200px;
}
.product-img {
    width: 100%;
    max-width: 150px;
}

.addCart-block {

}
.addQty_button {
    display: inline-block;
    width: 50px;
    height: 26px;
    border-right: 0;
    border: 1px solid #dbdbdb;
    outline: none;
	text-align: center;
}
.addCart_button {
    display: inline-block;
    width: 100px;
    height: 30px;
    margin: 0;
    vertical-align: middle;
    font-size: 16px;
    font-weight: normal;
    color: #fff;
    text-align: center;
    border-left: 0;
    border: 3px solid #0d897f;
    border-radius: 0 4px 4px 0;
    background-color: #0d897f;
    outline: none;
    transition: all 0.5s ease-in;
}

.addCart_button:hover, .addCart_button:active {
    color: #000;
    background: #fff;
}

.addCart_button a {
    color: #0d897f;
    text-decoration: none;
}

.inCart {
    background-color: #fff;
    color: #0d897f;
    border: 3px solid #0d897f;
    font-size: 14px;
    font-weight: bold;
}



@media (max-width: 1100px) {     
	
    .product-wrap {
        width: 48%;
    }
	
}

@media (max-width: 860px) {
    .product-wrap {
        width: 30%;
    }
    .product-title {
        height: 60px;
        font-size: 16px;
        line-height: 16px;
    }
}

@media (max-width: 780px) {
    .product-wrap {
        width: 48%;
    }
}
@media (max-width: 600px) {
    .products-container {
        justify-content: center;
    }
    .product-wrap {
        width: 80%;
    }
}

@media (max-width: 340px) {
}
