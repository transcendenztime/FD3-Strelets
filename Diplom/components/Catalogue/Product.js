"use strict";

import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';

import {connect} from 'react-redux';
import { prod_add, sum_qty } from '../../redux/cartAC';

import { setLocalStorage } from '../../services/LocalStorage';
import { round2 } from '../../services/round2';

import AddSpinner from './AddSpinner'

import { convertLink } from '../../services/LinkConverter';

import '../../styles/Catalogue/Product.css';

class Product extends React.PureComponent {
  
  static propTypes = { 
    info: PropTypes.shape({
       id: PropTypes.number.isRequired,
       name: PropTypes.string.isRequired,
       brand: PropTypes.string.isRequired,
       imgUrl: PropTypes.string.isRequired,
       price: PropTypes.number.isRequired,
       category: PropTypes.string.isRequired,
       description: PropTypes.string.isRequired,
      }),
  };

  state = {
    info: this.props.info,
    inCart:0, //1 - в корзине
  };
  
  newQTYRef = null;

  setNewQTYRef = (ref) => {
    this.newQTYRef = ref
  }

  addIntoCart = () => {
    let qty = +this.newQTYRef.value;
    let sum = round2(qty * this.props.info.price, 100); // округляем до вида 0,00
    let info = {...this.state.info, qty: qty, sum: sum};

    this.props.dispatch( prod_add(this.state.info.id, info) ); //добавляем в корзину
    this.props.dispatch( sum_qty(qty) ); //добавляем количество продуктов в корзине для отображения в header

    setLocalStorage(this.state.info.id, info);

    this.setState( {inCart:1} );
  }
  
  render() {
    let link = convertLink(this.props.info.category);

    return (
        <div className="product-wrap">
          <h3>
            <NavLink to = {"/catalogue/"+ link + "/" + this.props.info.id} className = "product-title">
              {this.props.info.name}
            </NavLink>
          </h3>
          <NavLink to = {"/catalogue/"+ link + "/" + this.props.info.id} className = "product-title product-img-container">
            <img src = {this.props.info.imgUrl} className = "product-img" />
          </NavLink>
          <p>{this.props.info.price + " руб."}</p>
          <div className = "addCart-block">
            <input type='number' step = "1" min = "1" defaultValue = {1} title = "Кол-во" ref = {this.setNewQTYRef} className = "addQty_button" />
            <button onClick = {this.addIntoCart} className = {this.state.inCart == 1? "addCart_button inCart" : "addCart_button"} title = "Добавить в корзину">
              {this.state.inCart != 1 ? 
                <i className="fas fa-shopping-cart"></i> 
                :<NavLink to = "/cart">Корзина</NavLink> 
              }              
            </button>
            <AddSpinner  class = {this.state.inCart !=1 ? 'invisibleAddSpinner':'visibleAddSpinner'}/>
          </div>
        </div>          
    );

  }

}

 
const mapStateToProps = function (state) {
  // этому компоненту ничего не нужно из хранилища Redux
  return { }; 
};

export default connect(mapStateToProps)(Product);
