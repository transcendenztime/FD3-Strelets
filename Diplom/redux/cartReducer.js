import {  PROD_ADD, PROD_DEL, CART_CLEAR, QTY_ADD, DELIVERY_CHNG } from './cartAC';

const initState={

  // ключ - идентификатор prod, значение - info
  products: {},
  delivery: 0,  
}


function cartReducer(state = initState, action) {
  switch (action.type) {
    
    case PROD_ADD: {
      let newProdInfo = action.info;
      
      //если в корзине уже есть данный товар, суммируем добавляемое с тем, что есть
      if (state.products[action.prodid]) {
        newProdInfo.qty += state.products[action.prodid].qty;
        newProdInfo.sum += state.products[action.prodid].sum;
      }

      let newState = {...state,
        products: {...state.products,
          [action.prodid]: newProdInfo}};
      return newState;
    }    

    case PROD_DEL: {
      let newState = {...state};
      delete newState.products[action.prodid];
      return newState;
    }    

    case CART_CLEAR: {
      //чистим корзину
      let newState = {...initState};
      return newState;
    }

    case QTY_ADD: {
      //меняем количество единиц товара и общую сумму денег за это количество единиц товара 
      let newState = {...state,
        products: {...state.products,
          [action.prodid]: {...state.products[action.prodid],
            qty: action.newQTY, sum: action.newSum }
        }
      };

      return newState;
    }
    
    case DELIVERY_CHNG: {
      let newState = {...state,
        delivery: action.num};
      return newState;
    } 
    
    default:
      return state;
  }
}

export default cartReducer;
