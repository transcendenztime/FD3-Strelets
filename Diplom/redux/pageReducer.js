import {  PAGE_CHANGE } from './pageAC';

const initState={
    page: 1,
}

function pageReducer(state = initState, action) {
  switch (action.type) {
    
    case PAGE_CHANGE: {
      
      //если страница та же, ничего не делаем
      if (state.page != action.page) {
        let newState = {...state,
          page: action.page};
        return newState;
      } else {return state}
      
    }    
    
    default:
      return state;
  }
}

export default pageReducer;
