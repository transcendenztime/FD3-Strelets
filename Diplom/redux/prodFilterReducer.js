import {  PROD_FILTER } from './prodFilterAC';

const initState={
    prodFilter: '',
}


function prodFilterReducer(state = initState, action) {
  switch (action.type) {
    
    case PROD_FILTER: {
            
      let newState = {...state,
        prodFilter: action.name};
        console.log(action.name)
      return newState;
      
    }    
    
    default:
      return state;
  }
}

export default prodFilterReducer;
