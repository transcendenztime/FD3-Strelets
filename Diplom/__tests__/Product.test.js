"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';

import { createStore } from 'redux';
import combinedReducer from '../redux/reducers.js';
import { Provider } from 'react-redux';

import Product from '../components/Catalogue/Product';

const prod = {
  "id": 1,
  "name": "Томат Гранатовая капля",
  "brand": "Гавриш",
  "category": "Семена овощей",
  "price": 1.32,
  "description": "Сладкий фруктовый вкус. Среднеспелый сорт для пленочных теплиц и открытого грунта.",
  "imgUrl": "./img/ed940bf6db9540b843b64bacabbae4ae.jpg",
  "qty": 1,
  "sum": 1.32
};
    

let store=createStore(combinedReducer);

test('внешний вид и работа Product', () => {

  // создаём тестовую версию компонента
  const component = renderer.create(
    <Router>
      <Provider store={store}>
        <Product info = {prod} />
      </Provider>
    </Router>
  );

  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

});
