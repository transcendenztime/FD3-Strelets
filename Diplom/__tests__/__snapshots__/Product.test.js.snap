// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`внешний вид и работа Product 1`] = `
<div
  className="product-wrap"
>
  <h3>
    <a
      aria-current={null}
      className="product-title"
      href="/catalogue/semena-ovoshchey/1"
      onClick={[Function]}
      style={undefined}
    >
      Томат Гранатовая капля
    </a>
  </h3>
  <a
    aria-current={null}
    className="product-title product-img-container"
    href="/catalogue/semena-ovoshchey/1"
    onClick={[Function]}
    style={undefined}
  >
    <img
      className="product-img"
      src="./img/ed940bf6db9540b843b64bacabbae4ae.jpg"
    />
  </a>
  <p>
    1.32 руб.
  </p>
  <div
    className="addCart-block"
  >
    <input
      className="addQty_button"
      defaultValue={1}
      min="1"
      step="1"
      title="Кол-во"
      type="number"
    />
    <button
      className="addCart_button"
      onClick={[Function]}
      title="Добавить в корзину"
    >
      <i
        className="fas fa-shopping-cart"
      />
    </button>
    <div
      className="invisibleAddSpinner"
    >
      <i
        className="fa fa-spinner fa-spin fa-fw"
      />
    </div>
  </div>
</div>
`;
